File: _algos/sorts/shell/shell.py
#!python3
"""
w1ld [dog] inbox dot ru
"""

n = int(input().strip())
a = [int(i) for i in input().strip().split(' ')]

gap = 1
while gap < n:
    gap = gap*3 + 1
gap //= 3

count = 0;
while gap > 0:
    for i in range(gap, n, 1):
        j = i
        while j-gap >= 0 and a[j-gap] > a[j]:
            a[j-gap], a[j] = a[j], a[j-gap]
            j -= gap
            count += 1
    gap //= 3

# print(' '.join(str(e) for e in a))
print('Exchanges:',count)

File: _algos/sorts/shell/shell.c
#include <stdio.h>
#include <stdlib.h>


int main() 
{
    int n;
    scanf("%d", &n);
    int a[n];
    int i;
    for (i = 0; i < n; ++i) {
        scanf("%d", (a+i));
    }

    int gap = 1;
    while (gap < n)
        gap = gap * 3 + 1;
    gap /= 3;

    long long count = 0;

    while (gap >= 1)
    {
        for (i = gap; i < n; i++) {
            int j = i;
            while (j-gap >= 0 && a[j-gap] > a[j])
            {
                int t = a[j-gap];
                a[j-gap] = a[j];
                a[j] = t;
                j -= gap;
                count++;
            }
        }
        gap /= 3;
    }

    /*for (int i = 0; i < n; i++)*/
        /*printf("%d ", a[i]);*/
    /*printf("\n");*/

    printf("Exchanges: %d\n", count);
    return 0;
}
